src/utils.cpp

```diff
@@ -332,8 +332,153 @@ void cnf_to_dimacs(Term cnf, std::ostringstream & y)
      y << 0;  // Symbolizing end of line
      y << "\n";
   }
+} 
+
+Term to_cnf(Term formula, SmtSolver s)
+{
+  
+  std::map<Term, Term>ma;//map ma contains the mapping of each term to it's new symbol in tseytin's transformation
+  
+  Sort boolsort = formula->get_sort();
+  assert(boolsort->get_sort_kind() == BOOL);
+  //vec stores the formulas which yet need to be given a symbolic name
+  TermVec vec = { formula };
+
+  //returning symbolic constants directly
+  if(formula->is_symbolic_const())
+  {
+    return formula;
+  }
+
+  static int pt = 1;// a pointer which will decide the next symbols name
+
+  auto term_gen = [&]
+  {
+    return (s->make_symbol("cnf__tseitin__fresh__var__" + std::to_string(pt++), boolsort));
+  };
+
+  auto give_symbolic_name=[&](Term t, bool push)
+  {//takes input as the term to be processed and a boolean variable push to specify if we need to push it back into our vector "vec"
+    if(t->is_symbolic_const())
+    {//symbolic constants are not given new symbolic names
+      return t;
+    }
+    else
+    {
+      if(ma.find(t) != ma.end())
+      {// term has already been visited before, so just return it's name mapped earlier and don't push it back in the vector
+        return ma[t];
+      }
+      else
+        {//term is being visited the first time, so give it a new symbolic name and push it back in the vector if the boolean variable says so 
+        ma[t] = term_gen();
+        if(push)
+        {
+          vec.push_back(t);
+        }
+        return ma[t];
+      }
+    }
+  };
+  
+  //reduced stores (c) <-> (a op b). (c) as the first term of the pair and (a op b) as the second term in the pair
+  std::vector<std::pair<Term, Term>>reduced;
+  
+  while(!vec.empty())
+  {
+    Term u = vec.back();
+    vec.pop_back();
+    smt::Op op = u->get_op();
+    if(op == smt::And || op == smt::Or || op == smt::Xor || op == smt::Implies)
```

Comment:
I could be missing if this is handled elsewhere, but you might have one other case: `Equal`. There's no `iff` in SMT-LIB, instead it's represented as an equality between booleans. E.g., `a <-> b` is actually `a = b`. So you might have to check if you have an equality between two booleans.

Reply:
Thanks! I had missed this